### ISSUE TITLE
Language support

### ISSUE THREAD

**Issue by cajoek** [cdqa-suite/cdQA#236]
Just wondering if it is possible to use cdQA with other languages than English, directly or with some modifications?

Thank you!

**Comment by fmikaelian** [comment]
Hi @cajoek 

It is possible to achieve that by changing both the reader and retriever of cdQA (see our [blog explanations](https://towardsdatascience.com/how-to-create-your-own-question-answering-system-easily-with-python-2ef8abc8eb5)).

The cdQA reader currently uses the english version of BERT, but there is a pre-trained multi-language version of BERT and other specific language versions of BERT that could be used. See the full pre-trained BERT models list below:

https://github.com/huggingface/pytorch-transformers/blob/7060766490240f7f1a63dce4c1ca6d0abfd8555d/pytorch_transformers/modeling_bert.py#L52-L66

You will need to fine-tune the model for QA using a custom language QA dataset, starting from the pre-trained multi-language weights mentionned above. For instance, that would work easily for Korean since there is a SQuAD-like dataset in Korean already available (https://korquad.github.io/KorQuad%201.0/)

Here is the raw code to train a reader on a new language:

```python
from cdqa.reader.bertqa_sklearn import BertProcessor, BertQA

train_processor = BertProcessor(bert_model='bert-base-multilingual-uncased', do_lower_case=True, is_training=True)
train_examples, train_features = train_processor.fit_transform(X='your-squad-like-dataset.json')

reader = BertQA(bert_model='bert-base-multilingual-uncased',
                train_batch_size=12,
                learning_rate=3e-5,
                num_train_epochs=2,
                do_lower_case=True,
                output_dir='models')

reader.fit(X=(train_examples, train_features))
```

Finally, the cdQA retriever also has to be modified, using a tokenizer and stop-words adapted to your language.

You could then use those models to perform prediction on your custom language.

**Comment by cajoek** [comment]
Okay, thanks a lot! I will try it. If I want to use cdQA for Swedish then I guess  the only way is to create a custom QA dataset with the cdQA-annotator for fine tuning and find a tokenizer and stop-words.

**Comment by fmikaelian** [comment]
Hi @cajoek 

Yes, creating a swedish QA dataset would be the best way to use cdQA for swedish. Unfortunately, QA datasets are only available for a few language at the moment since it is costly to build those. We have the same issue with French 😅

Some recent techniques allow for automatic creation of English question-answer pair (https://github.com/facebookresearch/UnsupervisedQA). It might be interesting to follow-up these.

**Comment by cajoek** [comment]
Hi again!
If you or anyone else want to know this is the code I ended up with:
```
import pandas as pd

from cdqa.utils.filters import filter_paragraphs
from cdqa.utils.converters import df2squad
from cdqa.reader.bertqa_sklearn import BertProcessor, BertQA
from cdqa.retriever.tfidf_sklearn import TfidfRetriever
from cdqa.pipeline.cdqa_sklearn import QAPipeline
from spacy.lang.sv import Swedish
from spacy.lang.sv.stop_words import STOP_WORDS


def get_reader(dataset_path='data/cdqa-id-v1.1.json'):
    train_processor = BertProcessor(bert_model='bert-base-multilingual-uncased', do_lower_case=True, is_training=True)
    train_examples, train_features = train_processor.fit_transform(X=dataset_path)

    reader = BertQA(bert_model='bert-base-multilingual-uncased',
                    train_batch_size=12,
                    learning_rate=3e-5,
                    num_train_epochs=2,
                    warmup_proportion=0.0,
                    do_lower_case=True,
                    output_dir='models')
    reader.fit(X=(train_examples, train_features))
    return reader


def get_retriever():
    nlp = Swedish()
    def tokenizer(text):
        return nlp(text).text.split(' ')

    retriever = TfidfRetriever(
        tokenizer=tokenizer,
        stop_words=STOP_WORDS
        )
    return retriever


def df_from_dir(dir_path):
    all_files = glob.glob(dir_path + "/*.txt")
    data = []
    for filename in all_files:
        paragraphs = [line.rstrip('\n') for line in open(filename)]
        data.append({
            'title': splitext(basename(filename))[0],
            'paragraphs': paragraphs
        })
    df = pd.DataFrame(data)
    return df


if __name__ == "__main__":
    df = df_from_dir('data/se/')

    retriever = get_retriever()
    reader = get_reader()

    cdqa_pipeline = QAPipeline(reader=reader)
    cdqa_pipeline.retriever = retriever
    cdqa_pipeline.fit_retriever(X=df)

    query = '.'
    while query != '':
        query = input('Query: ')
        prediction = cdqa_pipeline.predict(X=query)

        print('query: {}'.format(query))
        print('answer: {}'.format(prediction[0]))
        print('title: {}'.format(prediction[1]))
        print('paragraph: {}'.format(prediction[2]))
```
Does it look like what you expected? The performance is really bad but that could just be because my custom Swedish dataset is tiny.

##### Extra Things i had to do:
 - `warmup_proportion=0.0` otherwise I got the error: `Training beyond specified 't_total'. Learning rate multiplier set to 0.0. Please set 't_total' of WarmupLinearSchedule correctly.`
 - Create a json with `df2squad` and then used the cdQA-annotator to annotate it and make it into a dataset
 - Manually add random ids in the json to the qas

**Comment by fmikaelian** [comment]
Hi @cajoek 

Thanks for sharing you code, very interesting! 

Concerning the performance, how big is your QA dataset? As an example, KorQuAD 1.0 has 70,000+ question-answer pairs and SQuAD2.0 has about 150,000. How did you evaluate the pipeline performance?

Also, you can probably check for performance of the retriever vs. reader. Is the retriever working as expected?

Concerning the random ids during annotation, it is a missing feature we need to add to the annotator (see https://github.com/cdqa-suite/cdQA-annotator/issues/29) I'll increase its priority.

Hope it helps!

**Comment by cajoek** [comment]
Hello @fmikaelian 
The dataset is really tiny only about 100 question-answer pairs so I'm not expecting a good performance. But if I could build up a larger Swedish dataset over a longer time I want to know that I can use it.

**Comment by andrelmfarias** [comment]
Hi @cajoek,

A possible idea would be to translate SQuAD to your language and then trying training the model to have a first hint of its performance. I don't know if the translation models are performant in Swedish though...

**Comment by andrelmfarias** [comment]
Please also be aware that the performance of the system is also very dependent on the Retriever. I see in your code that you define a specific tokenizer for Swedish, which is a good first step, but I think you also might need to fine-tune some hyperparameters (stop_words, min_df, max_df, etc..).
